# A guided project on Dataquest

* The project is about to help build an app that generates maximum profits.

* The goal of this project is to attract more users who could interact with ads on our free app

Open two data_sets **`AppleStore.csv`** and **`googleplaystore.csv`**

In [2]:
from csv import reader
apple_data = open('AppleStore.csv')
google_data = open('googleplaystore.csv')

data_apple = reader(apple_data)
apple_data_list = list(data_apple)
apple_header = apple_data_list[0]

data_google = reader(google_data)
google_data_list = list(data_google)
google_header = google_data_list[0]

In [3]:
def explore_data(data_set, first_index, last_index, rows_and_columns=False):
    data_list = data_set[first_index: last_index]
    for row in data_list:
        print(row, '\n');
        
    if rows_and_columns:
        print('Number of rows: ', len(data_list))
        print('Number of columns: ', len(data_list[0]))

Extract data from `AppleStore.csv`

In [4]:

print(apple_header, '\n')
explore_data(apple_data_list, 1, 5, True)
print('\n')

['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic'] 

['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1'] 

['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1'] 

['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 'Games', '38', '5', '18', '1'] 

['420009108', 'Temple Run', '65921024', 'USD', '0.0', '1724546', '3842', '4.5', '4.0', '1.6.2', '9+', 'Games', '40', '5', '1', '1'] 

Number of rows:  4
Number of columns:  16




Extract data from `googleplaystore.csv`

In [5]:
print(google_header)
explore_data(google_data_list, 1, 5, True)

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']
['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up'] 

['Coloring book moana', 'ART_AND_DESIGN', '3.9', '967', '14M', '500,000+', 'Free', '0', 'Everyone', 'Art & Design;Pretend Play', 'January 15, 2018', '2.0.0', '4.0.3 and up'] 

['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up'] 

['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up'] 

Number of rows:  4
Number of columns:  13


## Data cleaning

- removing or correcting wrong data
- removing duplicate data
- modifying the data to fit the purpose of our analysis

In [6]:
print(google_header)
print(google_data_list[10473])

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']
['Life Made WI-Fi Touchscreen Photo Frame', '1.9', '19', '3.0M', '1,000+', 'Free', '0', 'Everyone', '', 'February 11, 2018', '1.0.19', '4.0 and up']


 10473 row has a missing column
 - delete the row
 - only to run the del state once, not more than once or twice, otherwise, it would delete more rows

In [7]:
del google_data_list[10473]

In [8]:
print(google_data_list[10473])

['osmino Wi-Fi: free WiFi', 'TOOLS', '4.2', '134203', '4.1M', '10,000,000+', 'Free', '0', 'Everyone', 'Tools', 'August 7, 2018', '6.06.14', '4.4 and up']


Search for other errors
- `9149` row
- with `NaN` values

In [9]:
print(google_data_list[9149])

['Command & Conquer: Rivals', 'FAMILY', 'NaN', '0', 'Varies with device', '0', 'NaN', '0', 'Everyone 10+', 'Strategy', 'June 28, 2018', 'Varies with device', 'Varies with device']


In [10]:
del google_data_list[9149]

In [11]:
print(google_data_list[9149])

['Star Wars™: Galaxy of Heroes', 'FAMILY', '4.5', '1461698', '67M', '10,000,000+', 'Free', '0', 'Everyone 10+', 'Role Playing', 'May 21, 2018', '0.12.334385', '4.1 and up']


## Removing Duplicate Entries

In [12]:
for app in google_data_list[1:]:
    name = app[0]
    if name == 'Instagram':
        print(name)
    
    if name == 'Facebook':
        print(name)
        
    if name == 'Twitter':
        print(name)

Facebook
Instagram
Instagram
Instagram
Twitter
Twitter
Instagram
Facebook
Twitter


Duplicate rows are noticeable!

Task
- Remove Duplicate rows

In [13]:
# creating two lists
# one with unique items
# one with duplicate items

unique_items = []
duplicate_items = []

for app_row in google_data_list[1:]:
    name = app_row[0]
    if not name in unique_items:
        unique_items.append(name)
    else:
        duplicate_items.append(name)
        
print('Unique list: \n')
print(unique_items[:10], '\n')
    
print('Number of duplicate apps: ', len(duplicate_items))
print('Example of duplicate apps: \n')
print(duplicate_items)

Unique list: 

['Photo Editor & Candy Camera & Grid & ScrapBook', 'Coloring book moana', 'U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'Sketch - Draw & Paint', 'Pixel Draw - Number Art Coloring Book', 'Paper flowers instructions', 'Smoke Effect Photo Maker - Smoke Editor', 'Infinite Painter', 'Garden Coloring Book', 'Kids Paint Free - Drawing Fun'] 

Number of duplicate apps:  1181
Example of duplicate apps: 

['Quick PDF Scanner + OCR FREE', 'Box', 'Google My Business', 'ZOOM Cloud Meetings', 'join.me - Simple Meetings', 'Box', 'Zenefits', 'Google Ads', 'Google My Business', 'Slack', 'FreshBooks Classic', 'Insightly CRM', 'QuickBooks Accounting: Invoicing & Expenses', 'HipChat - Chat Built for Teams', 'Xero Accounting Software', 'MailChimp - Email, Marketing Automation', 'Crew - Free Messaging and Scheduling', 'Asana: organize team projects', 'Google Analytics', 'AdWords Express', 'Accounting App - Zoho Books', 'Invoice & Time Tracking - Zoho', 'join.me - Simple Meetings', 'In

We only entered the first value of the data set that is encountered, but, do not know which value should be kept among the duplicates. 

To decide for this, we use, 

- a dictionary to keep track of the **`reviews`** of the highest frequency as a criterion

- We keep track of the records that are updated the last time. In this regard, the record with the highest review should be the one of our main focus for the latest record.

In [14]:
max_review = {}

for row in google_data_list[1:]:
    app_name = row[0]
    app_review = float(row[3])
    
    if app_name in max_review and max_review[app_name] < app_review:
        max_review[app_name] = app_review
        
    elif app_name not in max_review:
        max_review[app_name] = app_review
        

In [15]:
unique_google_list = []
duplicate_google_list = []

for row in google_data_list[1:]:
    name = row[0]
    total_app_reviews = float(row[3])
    
    if (max_review[name] == total_app_reviews) and row not in duplicate_google_list:
        unique_google_list.append(row)
    else:
        duplicate_google_list.append(row)

        
print(unique_google_list[:5])

[['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up'], ['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up'], ['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up'], ['Pixel Draw - Number Art Coloring Book', 'ART_AND_DESIGN', '4.3', '967', '2.8M', '100,000+', 'Free', '0', 'Everyone', 'Art & Design;Creativity', 'June 20, 2018', '1.1', '4.4 and up'], ['Paper flowers instructions', 'ART_AND_DESIGN', '4.4', '167', '5.6M', '50,000+', 'Free', '0', 'Everyone', 'Art & Design', 'March 26, 2017', '1.0', '2.3 and up']]


In [16]:
for row in unique_google_list:
    if row[0] == 'Instagram':
        print(row)
        

['Instagram', 'SOCIAL', '4.5', '66577446', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']


In [17]:
for row in duplicate_google_list:
    if row[0] == 'Instagram':
        print(row)

['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66509917', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']


The `ln[16]` and `ln[17]` confirms that the greatest reviews is included in the original unique list, i.e., **`unique_google_list`**

In [18]:
google_data_list[7930:7945]

[['Firefighters First CU',
  'FINANCE',
  '4.8',
  '495',
  '12M',
  '5,000+',
  'Free',
  '0',
  'Everyone',
  'Finance',
  'May 1, 2018',
  '5.7.1.0',
  '5.0 and up'],
 ['MYCU TX Mobile Banking',
  'FINANCE',
  '3.0',
  '4',
  'Varies with device',
  '100+',
  'Free',
  '0',
  'Everyone',
  'Finance',
  'August 1, 2018',
  '1.0.2',
  'Varies with device'],
 ['Amazing Fart Sounds',
  'FAMILY',
  '4.0',
  '28136',
  'Varies with device',
  '5,000,000+',
  'Free',
  '0',
  'Everyone',
  'Casual',
  'August 24, 2017',
  'Varies with device',
  'Varies with device'],
 ['Navigator PRO - GPS Navigation with Offline Maps',
  'TRAVEL_AND_LOCAL',
  '3.6',
  '892',
  '70M',
  '50,000+',
  'Paid',
  '$2.49',
  'Everyone',
  'Travel & Local',
  'August 1, 2018',
  '1.08',
  '4.0 and up'],
 ['My Vodafone (GR)',
  'COMMUNICATION',
  '3.8',
  '12667',
  '34M',
  '1,000,000+',
  'Free',
  '0',
  'Everyone',
  'Communication',
  'August 1, 2018',
  '4.5.5.1-2090AL-REL',
  '4.3 and up'],
 ['HTC Transfe

In [19]:
def is_english(strings):

    for character in strings:

        if ord(character) > 127:
            return False
        
    return True


In [20]:
is_english('爱奇艺PPS -《欢乐颂2》电视剧热播')
           

False

In [21]:
is_english('Instagram')

True

In [22]:
is_english('Docs To Go™ Free Office Suite')

False

In [23]:
is_english('Instachat 😜')

False

In [24]:
print(ord('😜'))

128540


In [25]:
print(ord('™'))

8482


Here ™ and 😜, emoji or special charater are supposed 
to make our function show False result even for some English inputs, hence, 
we shall be ignoring up to 3 special characters 
that have ascii values more than 127.

In [26]:
def is_english(strings):
    count = 0

    for character in strings:
        
        if ord(character) > 127 and count > 3:
            return False
        
        if ord(character) > 127:
            count += 1
        
    return True

In [27]:
print(is_english('Instachat 😜'))

True


In [28]:
print(is_english('Docs To Go™ Free Office Suite'))

True


In [29]:
print(is_english('爱奇艺PPS -《欢乐颂2》电视剧热播'
))

False


The function is working fine now.

In [30]:
android_apps = []
count_google_non_english = 0

for row in unique_google_list:
    name = row[0]
    
    if not is_english(name):
        print(name)
        count_google_non_english += 1
        continue
        
    android_apps.append(row)

   

print('\n\n')
print('Non English for google ::', count_google_non_english)

Flame - درب عقلك يوميا
သိင်္ Astrology - Min Thein Kha BayDin
РИА Новости
صور حرف H
L.POINT - 엘포인트 [ 포인트, 멤버십, 적립, 사용, 모바일 카드, 쿠폰, 롯데]
RMEduS - 음성인식을 활용한 R 프로그래밍 실습 시스템
AJ렌터카 법인 카셰어링
Al Quran Free - القرآن (Islam)
中国語 AQリスニング
Ay Yıldız Duvar Kağıtları
বাংলা টিভি প্রো BD Bangla TV
Cъновник BG
CSCS BG (в български)
뽕티비 - 개인방송, 인터넷방송, BJ방송
BL 女性向け恋愛ゲーム◆俺プリクロス
SecondSecret ‐「恋を読む」BLノベルゲーム‐
BL 女性向け恋愛ゲーム◆ごくメン
あなカレ【BL】無料ゲーム
감성학원 BL 첫사랑
BQ-መጽሐፍ ቅዱሳዊ ጥያቄዎች
BS Calendar / Patro / पात्रो
Vip视频免费看-BT磁力搜索
Билеты ПДД CD 2019 PRO
Bonjour 2017 Abidjan CI ❤❤❤❤❤
الفاتحون Conquerors
DG ग्राम / Digital Gram Panchayat
DM הפקות
DW فارسی By dw-arab.com
لعبة تقدر تربح DZ
বাংলাflix
RPG ブレイジング ソウルズ アクセレイト
ECナビ×シュフー
أحداث وحقائق | خبر عاجل في اخبار العالم
EG SIM CARD (EGSIMCARD, 이지심카드)
パーリーゲイツ公式通販｜EJ STYLE（イージェイスタイル）
FAHREDDİN er-RÂZİ TEFSİRİ
I'm Rich/Eu sou Rico/أنا غني/我很有錢
Ey Sey Storytime រឿងនិទានតាឥសី
哈哈姆特不EY
FP Разбитый дисплей



Non English for google :: 40


In [31]:
ios_apps = []
count_ios_non_english = 0

for row in apple_data_list[1:]:
    name = row[0]
    
    if not is_english(name):
        print(name)
        count_ios_non_english += 1
        continue
    
    ios_apps.append(row)
        

print('\n\n')
print('Non English for Apple :: ', count_ios_non_english)




Non English for Apple ::  0


In [32]:
android_apps[:5]

[['Photo Editor & Candy Camera & Grid & ScrapBook',
  'ART_AND_DESIGN',
  '4.1',
  '159',
  '19M',
  '10,000+',
  'Free',
  '0',
  'Everyone',
  'Art & Design',
  'January 7, 2018',
  '1.0.0',
  '4.0.3 and up'],
 ['U Launcher Lite – FREE Live Cool Themes, Hide Apps',
  'ART_AND_DESIGN',
  '4.7',
  '87510',
  '8.7M',
  '5,000,000+',
  'Free',
  '0',
  'Everyone',
  'Art & Design',
  'August 1, 2018',
  '1.2.4',
  '4.0.3 and up'],
 ['Sketch - Draw & Paint',
  'ART_AND_DESIGN',
  '4.5',
  '215644',
  '25M',
  '50,000,000+',
  'Free',
  '0',
  'Teen',
  'Art & Design',
  'June 8, 2018',
  'Varies with device',
  '4.2 and up'],
 ['Pixel Draw - Number Art Coloring Book',
  'ART_AND_DESIGN',
  '4.3',
  '967',
  '2.8M',
  '100,000+',
  'Free',
  '0',
  'Everyone',
  'Art & Design;Creativity',
  'June 20, 2018',
  '1.1',
  '4.4 and up'],
 ['Paper flowers instructions',
  'ART_AND_DESIGN',
  '4.4',
  '167',
  '5.6M',
  '50,000+',
  'Free',
  '0',
  'Everyone',
  'Art & Design',
  'March 26, 2017

In [33]:
ios_apps[:5]

[['284882215',
  'Facebook',
  '389879808',
  'USD',
  '0.0',
  '2974676',
  '212',
  '3.5',
  '3.5',
  '95.0',
  '4+',
  'Social Networking',
  '37',
  '1',
  '29',
  '1'],
 ['389801252',
  'Instagram',
  '113954816',
  'USD',
  '0.0',
  '2161558',
  '1289',
  '4.5',
  '4.0',
  '10.23',
  '12+',
  'Photo & Video',
  '37',
  '0',
  '29',
  '1'],
 ['529479190',
  'Clash of Clans',
  '116476928',
  'USD',
  '0.0',
  '2130805',
  '579',
  '4.5',
  '4.5',
  '9.24.12',
  '9+',
  'Games',
  '38',
  '5',
  '18',
  '1'],
 ['420009108',
  'Temple Run',
  '65921024',
  'USD',
  '0.0',
  '1724546',
  '3842',
  '4.5',
  '4.0',
  '1.6.2',
  '9+',
  'Games',
  '40',
  '5',
  '1',
  '1'],
 ['284035177',
  'Pandora - Music & Radio',
  '130242560',
  'USD',
  '0.0',
  '1126879',
  '3594',
  '4.0',
  '4.5',
  '8.4.1',
  '12+',
  'Music',
  '37',
  '4',
  '1',
  '1']]

In [34]:
explore_data(android_apps, 0, 3, True)

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up'] 

['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up'] 

['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up'] 

Number of rows:  3
Number of columns:  13


## Isolating the Free Apps

Till now we have done:

- Removed inaccurate fillings of data or nonfilling

- Removed Duplicate app entries

- Removed non-English apps

Now we are to isolate the Free Apps that is what we need for our purpose. 
This would be the last step in the data cleaning process.

In [35]:
android_apps[:3]

[['Photo Editor & Candy Camera & Grid & ScrapBook',
  'ART_AND_DESIGN',
  '4.1',
  '159',
  '19M',
  '10,000+',
  'Free',
  '0',
  'Everyone',
  'Art & Design',
  'January 7, 2018',
  '1.0.0',
  '4.0.3 and up'],
 ['U Launcher Lite – FREE Live Cool Themes, Hide Apps',
  'ART_AND_DESIGN',
  '4.7',
  '87510',
  '8.7M',
  '5,000,000+',
  'Free',
  '0',
  'Everyone',
  'Art & Design',
  'August 1, 2018',
  '1.2.4',
  '4.0.3 and up'],
 ['Sketch - Draw & Paint',
  'ART_AND_DESIGN',
  '4.5',
  '215644',
  '25M',
  '50,000,000+',
  'Free',
  '0',
  'Teen',
  'Art & Design',
  'June 8, 2018',
  'Varies with device',
  '4.2 and up']]

In [36]:
free_google_final_dataset = []
free_apple_final_dataset = []

remaining_android_apps = 0
remaining_ios_apps = 0

for row in android_apps:
    google_price = row[7]
    if google_price == '0':
        free_google_final_dataset.append(row)
        remaining_android_apps += 1
        
    
    
for row in ios_apps:
    ios_price = row[4]
    if ios_price == '0.0':
        free_apple_final_dataset.append(row)
        remaining_ios_apps += 1

print('Remaining android apps :: ', remaining_android_apps)
print('Remaining ios apps :: ', remaining_ios_apps)
    

Remaining android apps ::  9230
Remaining ios apps ::  4056


In [37]:
explore_data(free_google_final_dataset, 0, 3, True)
explore_data(free_apple_final_dataset, 0, 3, True)

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up'] 

['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up'] 

['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up'] 

Number of rows:  3
Number of columns:  13
['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1'] 

['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1'] 

['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 'Ga

## Genre frequency

- We are going to see which genre is popular to decide on building app on that genre

- we will create a frequency dictionary for genres that will help us decide

In [42]:
android_genre_frequency = {}

for row in free_google_final_dataset:
    
    genre = row[9]
    
    if genre in android_genre_frequency:
        android_genre_frequency[genre] += 1
    elif genre not in android_genre_frequency:
        android_genre_frequency[genre] = 1


print(genre_frequency)

{'Art & Design': 53, 'Art & Design;Creativity': 6, 'Auto & Vehicles': 82, 'Beauty': 53, 'Books & Reference': 192, 'Business': 436, 'Comics': 54, 'Comics;Creativity': 1, 'Communication': 299, 'Dating': 196, 'Education': 494, 'Education;Creativity': 4, 'Education;Education': 34, 'Education;Pretend Play': 8, 'Education;Brain Games': 3, 'Entertainment': 563, 'Entertainment;Brain Games': 7, 'Entertainment;Creativity': 3, 'Entertainment;Music & Video': 15, 'Events': 63, 'Finance': 332, 'Food & Drink': 113, 'Health & Fitness': 304, 'House & Home': 81, 'Libraries & Demo': 83, 'Lifestyle': 355, 'Lifestyle;Pretend Play': 1, 'Card': 40, 'Arcade': 165, 'Puzzle': 103, 'Racing': 88, 'Sports': 328, 'Casual': 157, 'Simulation': 182, 'Adventure': 60, 'Trivia': 37, 'Action': 278, 'Word': 23, 'Role Playing': 83, 'Strategy': 81, 'Board': 34, 'Music': 19, 'Action;Action & Adventure': 10, 'Casual;Brain Games': 13, 'Educational;Creativity': 3, 'Puzzle;Brain Games': 16, 'Educational;Education': 38, 'Casual;Pr

In [44]:
apple_genre_frequency = {}

for row in free_apple_final_dataset:
    genre = row[11]
    
    if genre in apple_genre_frequency:
        apple_genre_frequency[genre] += 1
    elif genre not in apple_genre_frequency:
        apple_genre_frequency[genre] = 1
        
        
print(apple_genre_frequency)

{'Social Networking': 143, 'Photo & Video': 167, 'Games': 2257, 'Music': 67, 'Reference': 20, 'Health & Fitness': 76, 'Weather': 31, 'Utilities': 109, 'Travel': 56, 'Shopping': 121, 'News': 58, 'Navigation': 20, 'Lifestyle': 94, 'Entertainment': 334, 'Food & Drink': 43, 'Sports': 79, 'Book': 66, 'Finance': 84, 'Education': 132, 'Productivity': 62, 'Business': 20, 'Catalogs': 9, 'Medical': 8}


- Frequency table in ascending order

- make the function usable for all columns

- represent the values in percentage


In [66]:
def freq_table(dataset, index):
    table = {}
#     table for frequency
    total = 0
#     total app
    
    for row in dataset:
        total += 1
        column_value = row[index]
        
        if column_value in table:
            table[column_value] += 1
            
        else:
            table[column_value] = 1
            
        table_percentages = {}
        
    for key in table:
        percentage = (table[key] / total) * 100
        table_percentages[key] = percentage

    return table_percentages

## sorting

In [67]:
def display_table(dataset, index):
    table = freq_table(dataset, index)
    table_display = []
    
    for key in table:
        key_val_as_tuple = (table[key], key)
        table_display.append(key_val_as_tuple)
#         appending tuple, from dictionary
        
    table_sorted = sorted(table_display, reverse = True)
    
    for entry in table_sorted:
        print(entry[1], ':', entry[0])

## Most common apps by Genre

In [78]:
display_table(free_google_final_dataset, 1)

# based on category android

FAMILY : 18.429035752979413
GAME : 9.469122426868907
TOOLS : 8.136511375947995
BUSINESS : 4.723726977248104
PRODUCTIVITY : 3.8894907908992415
LIFESTYLE : 3.856988082340195
MEDICAL : 3.71614301191766
FINANCE : 3.596966413867822
SPORTS : 3.4777898158179847
HEALTH_AND_FITNESS : 3.2936078006500544
COMMUNICATION : 3.2394366197183095
PERSONALIZATION : 3.217768147345612
PHOTOGRAPHY : 2.903575297941495
NEWS_AND_MAGAZINES : 2.8710725893824485
SOCIAL : 2.697724810400867
SHOPPING : 2.513542795232936
TRAVEL_AND_LOCAL : 2.426868905742145
DATING : 2.1235102925243767
BOOKS_AND_REFERENCE : 2.080173347778982
VIDEO_PLAYERS : 1.722643553629469
MAPS_AND_NAVIGATION : 1.3542795232936078
EDUCATION : 1.3217768147345612
FOOD_AND_DRINK : 1.2242686890574215
ENTERTAINMENT : 1.180931744312026
LIBRARIES_AND_DEMO : 0.8992416034669555
AUTO_AND_VEHICLES : 0.8884073672806068
HOUSE_AND_HOME : 0.8775731310942579
WEATHER : 0.7692307692307693
EVENTS : 0.6825568797399784
PARENTING : 0.628385698808234
ART_AND_DESIGN : 0.6175

From the Google dataset, when we see through the category column we can have certain findings as below:

---

- `FAMILY` is the category with highest percentage of 18.43.

- After then, `GAME` with 9.47, `TOOLS` with 8.14 that are seen a bit high than other category preferred

- Yet, `BUSINESS` 4.7%, `PRODUCTIVITY` with 3.89, and so on along with `LIFESTYLE`, `MEDICAL`, `FINANCE`, `SPORTS` etc are preferred. 

- **lowest** is the `BEAUTY` category with 0.57%

**Conclusion**

- Games and entertainment apps are the common that are preferred by the android users!

In [69]:
display_table(free_apple_final_dataset, -5)

# based on prime genre ios

Games : 55.64595660749507
Entertainment : 8.234714003944774
Photo & Video : 4.117357001972387
Social Networking : 3.5256410256410255
Education : 3.2544378698224854
Shopping : 2.983234714003945
Utilities : 2.687376725838264
Lifestyle : 2.3175542406311638
Finance : 2.0710059171597637
Sports : 1.947731755424063
Health & Fitness : 1.8737672583826428
Music : 1.6518737672583828
Book : 1.6272189349112427
Productivity : 1.5285996055226825
News : 1.4299802761341223
Travel : 1.3806706114398422
Food & Drink : 1.0601577909270217
Weather : 0.7642998027613412
Reference : 0.4930966469428008
Navigation : 0.4930966469428008
Business : 0.4930966469428008
Catalogs : 0.22189349112426035
Medical : 0.19723865877712032


Based on prime_genre of ios dataset column, it can be noticed that **`Games`** is the most preferred choice for basic users with 55.64% of preference of the total apps from the ios app store.

---

**Findings**

- `Games` would be the first choice

- `Entertainment` comes as a second choice

- `Photo & vide`, `Social Networking`, `Education` and so on are then preferred in a declining terms.

**Conclusion**

- The games and entertainment genre would be more profitable to build an app for.

In [76]:
display_table(free_google_final_dataset, -4)
# based on genre android

Tools : 8.125677139761647
Entertainment : 6.09967497291441
Education : 5.352112676056338
Business : 4.723726977248104
Productivity : 3.8894907908992415
Lifestyle : 3.8461538461538463
Medical : 3.71614301191766
Finance : 3.596966413867822
Sports : 3.553629469122427
Health & Fitness : 3.2936078006500544
Communication : 3.2394366197183095
Personalization : 3.217768147345612
Action : 3.0119176598049835
Photography : 2.903575297941495
News & Magazines : 2.8710725893824485
Social : 2.697724810400867
Shopping : 2.513542795232936
Travel & Local : 2.4160346695557964
Dating : 2.1235102925243767
Books & Reference : 2.080173347778982
Simulation : 1.971830985915493
Arcade : 1.7876489707475622
Video Players & Editors : 1.7009750812567714
Casual : 1.7009750812567714
Maps & Navigation : 1.3542795232936078
Food & Drink : 1.2242686890574215
Puzzle : 1.1159263271939328
Racing : 0.9534127843986998
Role Playing : 0.8992416034669555
Libraries & Demo : 0.8992416034669555
Auto & Vehicles : 0.8884073672806068


If we again see the google app store for the genre column, we can have following findings: 

---

- `Tools` is the app with highest preference with 8.1257% of use

- `Entertainment` is in the second preference with 6.1%

- `Education`, `Business` and others are preferred after then. 


In [82]:
apple_header

['id',
 'track_name',
 'size_bytes',
 'currency',
 'price',
 'rating_count_tot',
 'rating_count_ver',
 'user_rating',
 'user_rating_ver',
 'ver',
 'cont_rating',
 'prime_genre',
 'sup_devices.num',
 'ipadSc_urls.num',
 'lang.num',
 'vpp_lic']

## Most popular apps by Genre on the ios app store

We are going to calculate the averate number of installs for google play apps, but, for ios apps, there is no column as such hence, we are going to use rating_count_tot column to determine the average numbers of user ratings per app genre on the app store.

In [83]:
genres_ios = freq_table(free_apple_final_dataset, -5)
# prime_genre column

for genre in genres_ios:
    total = 0
#     to count the numbers of ratings per genre

    len_genre = 0
    
    for app in free_apple_final_dataset:
        genre_app = app[-5]
        
        if genre_app == genre:
            n_ratings = float(app[5])
            total += n_ratings
            len_genre += 1
            
    avg_n_ratings = total / len_genre
    print(genre, ':', avg_n_ratings)

Social Networking : 53078.195804195806
Photo & Video : 27249.892215568863
Games : 18924.68896765618
Music : 56482.02985074627
Reference : 67447.9
Health & Fitness : 19952.315789473683
Weather : 47220.93548387097
Utilities : 14010.100917431193
Travel : 20216.01785714286
Shopping : 18746.677685950413
News : 15892.724137931034
Navigation : 25972.05
Lifestyle : 8978.308510638299
Entertainment : 10822.961077844311
Food & Drink : 20179.093023255813
Sports : 20128.974683544304
Book : 8498.333333333334
Finance : 13522.261904761905
Education : 6266.333333333333
Productivity : 19053.887096774193
Business : 6367.8
Catalogs : 1779.5555555555557
Medical : 459.75


The reference genre is with highest numbers of user reviews then ther comes social networking, music, weather, and photo & video apps with highest number of user reviews.

In [87]:
for row in free_apple_final_dataset:
    if row[-5] == 'Reference':
        print(row[1], ':', row[5])
    


Bible : 985920
Dictionary.com Dictionary & Thesaurus : 200047
Dictionary.com Dictionary & Thesaurus for iPad : 54175
Google Translate : 26786
Muslim Pro: Ramadan 2017 Prayer Times, Azan, Quran : 18418
New Furniture Mods - Pocket Wiki & Game Tools for Minecraft PC Edition : 17588
Merriam-Webster Dictionary : 16849
Night Sky : 12122
City Maps for Minecraft PE - The Best Maps for Minecraft Pocket Edition (MCPE) : 8535
LUCKY BLOCK MOD ™ for Minecraft PC Edition - The Best Pocket Wiki & Mods Installer Tools : 4693
GUNS MODS for Minecraft PC Edition - Mods Tools : 1497
Guides for Pokémon GO - Pokemon GO News and Cheats : 826
WWDC : 762
Horror Maps for Minecraft PE - Download The Scariest Maps for Minecraft Pocket Edition (MCPE) Free : 718
VPN Express : 14
Real Bike Traffic Rider Virtual Reality Glasses : 8
教えて!goo : 0
彩库宝典-【官方版】 : 0
Jishokun-Japanese English Dictionary & Translator : 0
無料で音楽や写真・カメラの裏技アプリ for iPhone7 : 0


In [89]:
for row in free_apple_final_dataset:
    if row[-5] == 'Social Networking':
        print(row[1], ':', row[5])

Facebook : 2974676
Pinterest : 1061624
Skype for iPhone : 373519
Messenger : 351466
Tumblr : 334293
WhatsApp Messenger : 287589
Kik : 260965
ooVoo – Free Video Call, Text and Voice : 177501
TextNow - Unlimited Text + Calls : 164963
Viber Messenger – Text & Call : 164249
Followers - Social Analytics For Instagram : 112778
MeetMe - Chat and Meet New People : 97072
We Heart It - Fashion, wallpapers, quotes, tattoos : 90414
InsTrack for Instagram - Analytics Plus More : 85535
Tango - Free Video Call, Voice and Chat : 75412
LinkedIn : 71856
Match™ - #1 Dating App. : 60659
Skype for iPad : 60163
POF - Best Dating App for Conversations : 52642
Timehop : 49510
Find My Family, Friends & iPhone - Life360 Locator : 43877
Whisper - Share, Express, Meet : 39819
Hangouts : 36404
LINE PLAY - Your Avatar World : 34677
WeChat : 34584
Badoo - Meet New People, Chat, Socialize. : 34428
Followers + for Instagram - Follower Analytics : 28633
GroupMe : 28260
Marco Polo Video Walkie Talkie : 27662
Miitomo : 2

`Reference` and `Social Networking` apps should be the priority to build.

## Most popular apps by Genre on Google Play

We calculate the genre popularity based on the numbers of installs (6th) coloumn of free_google_final_dataset


In [103]:
free_google_final_dataset[1][5]
# install column

'5,000,000+'

In [97]:
display_table(free_google_final_dataset, 5)

1,000,000+ : 16.034669555796317
100,000+ : 11.451787648970747
10,000,000+ : 11.126760563380282
10,000+ : 9.924160346695558
1,000+ : 8.147345612134345
5,000,000+ : 7.096424702058505
100+ : 6.717226435536294
500,000+ : 5.568797399783315
50,000+ : 4.615384615384616
5,000+ : 4.398699891657638
10+ : 3.412784398699892
500+ : 3.14192849404117
50,000,000+ : 2.3293607800650054
100,000,000+ : 2.307692307692308
50+ : 1.8526543878656554
5+ : 0.7583965330444203
1+ : 0.4875406283856988
500,000,000+ : 0.314192849404117
1,000,000,000+ : 0.27085590465872156
0+ : 0.04333694474539545


We can see that most values are open-ended(100+, 1000+ etc)

---

For instance, we don't know whether an app with 100,000+ installs has 100,000 installs, 200,000, or 350,000. However, we don't need very precise data for our purposes — we only want to find out which app genres attract the most users.

We're going to leave the numbers as they are, which means that we'll consider that an app with 100,000+ installs has 100,000 installs, and an app with 1,000,000+ installs has 1,000,000 installs, and so on. To perform computations, however, we'll need to convert each install number from a string to a float. This means we need to remove the commas and the plus characters, or the conversion will fail and cause an error.

In [101]:
android_category = freq_table(free_google_final_dataset, 1)

android_category

{'ART_AND_DESIGN': 0.6175514626218852,
 'AUTO_AND_VEHICLES': 0.8884073672806068,
 'BEAUTY': 0.5742145178764897,
 'BOOKS_AND_REFERENCE': 2.080173347778982,
 'BUSINESS': 4.723726977248104,
 'COMICS': 0.5958829902491874,
 'COMMUNICATION': 3.2394366197183095,
 'DATING': 2.1235102925243767,
 'EDUCATION': 1.3217768147345612,
 'ENTERTAINMENT': 1.180931744312026,
 'EVENTS': 0.6825568797399784,
 'FINANCE': 3.596966413867822,
 'FOOD_AND_DRINK': 1.2242686890574215,
 'HEALTH_AND_FITNESS': 3.2936078006500544,
 'HOUSE_AND_HOME': 0.8775731310942579,
 'LIBRARIES_AND_DEMO': 0.8992416034669555,
 'LIFESTYLE': 3.856988082340195,
 'GAME': 9.469122426868907,
 'FAMILY': 18.429035752979413,
 'MEDICAL': 3.71614301191766,
 'SOCIAL': 2.697724810400867,
 'SHOPPING': 2.513542795232936,
 'PHOTOGRAPHY': 2.903575297941495,
 'SPORTS': 3.4777898158179847,
 'TRAVEL_AND_LOCAL': 2.426868905742145,
 'TOOLS': 8.136511375947995,
 'PERSONALIZATION': 3.217768147345612,
 'PRODUCTIVITY': 3.8894907908992415,
 'PARENTING': 0.62838

In [109]:
for category in android_category:
    total = 0
#     sum of installs specific to each genre

    len_category = 0
#     length of numbers of apps specific to each genre

    for row in free_google_final_dataset:
        category_app = row[1]
#         category column

        if category_app == category:
            number_of_installs = row[5]
#             installs column

            number_of_installs = number_of_installs.replace('+', '')
            number_of_installs = number_of_installs.replace(',', '')
            
            total += float(number_of_installs)
            len_category += 1
            
    avg_n_installs = total / len_category
    print(category, ':', avg_n_installs)

ART_AND_DESIGN : 1986335.0877192982
AUTO_AND_VEHICLES : 647317.8170731707
BEAUTY : 513151.88679245283
BOOKS_AND_REFERENCE : 8702574.270833334
BUSINESS : 1892091.9495412845
COMICS : 817657.2727272727
COMMUNICATION : 52882629.43478261
DATING : 854027.8418367347
EDUCATION : 2265163.9344262294
ENTERTAINMENT : 11105137.614678899
EVENTS : 253542.22222222222
FINANCE : 1374888.9518072288
FOOD_AND_DRINK : 2015387.1769911505
HEALTH_AND_FITNESS : 4261507.90131579
HOUSE_AND_HOME : 1737067.4197530865
LIBRARIES_AND_DEMO : 638503.734939759
LIFESTYLE : 1405377.6095505618
GAME : 17784747.654462244
FAMILY : 3760962.1928277486
MEDICAL : 141184.85131195336
SOCIAL : 24252457.437751003
SHOPPING : 8119993.900862069
PHOTOGRAPHY : 17952495.57835821
SPORTS : 3715983.436137072
TRAVEL_AND_LOCAL : 14204036.098214285
TOOLS : 10800324.199733688
PERSONALIZATION : 5990693.225589226
PRODUCTIVITY : 20319301.710306406
PARENTING : 542603.6206896552
WEATHER : 5074486.197183099
VIDEO_PLAYERS : 24727872.452830188
NEWS_AND_MA

From the above findings, 

- `Communication` is the highest installed apps for the user preference

- `Entertainment` and `Game` Catergory comes there after

Hence the trend is more in these categories for people to have more install records. Hence, our free app should be more installed reaching out to the market should we build an app related to these area

In [110]:
explore_data(free_google_final_dataset, 0, 3, True)

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up'] 

['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up'] 

['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up'] 

Number of rows:  3
Number of columns:  13


In [117]:
for apps in free_google_final_dataset:
    if apps[1] == 'COMMUNICATION':
        print(apps[0], ':', apps[5])

WhatsApp Messenger : 1,000,000,000+
Messenger for SMS : 10,000,000+
My Tele2 : 5,000,000+
imo beta free calls and text : 100,000,000+
Contacts : 50,000,000+
Call Free – Free Call : 5,000,000+
Web Browser & Explorer : 5,000,000+
Browser 4G : 10,000,000+
MegaFon Dashboard : 10,000,000+
ZenUI Dialer & Contacts : 10,000,000+
Cricket Visual Voicemail : 10,000,000+
TracFone My Account : 1,000,000+
Xperia Link™ : 10,000,000+
TouchPal Keyboard - Fun Emoji & Android Keyboard : 10,000,000+
Skype Lite - Free Video Call & Chat : 5,000,000+
My magenta : 1,000,000+
Android Messages : 100,000,000+
Google Duo - High Quality Video Calls : 500,000,000+
Seznam.cz : 1,000,000+
Antillean Gold Telegram (original version) : 100,000+
AT&T Visual Voicemail : 10,000,000+
GMX Mail : 10,000,000+
Omlet Chat : 10,000,000+
My Vodacom SA : 5,000,000+
Microsoft Edge : 5,000,000+
WhatsApp Messenger : 1,000,000,000+
Messenger – Text and Video Chat for Free : 1,000,000,000+
imo free video calls and chat : 500,000,000+
Ca

In [120]:
for apps in free_google_final_dataset:
    if apps[1] == 'ENTERTAINMENT':
        print(apps[0], ':', apps[5])

Complete Spanish Movies : 1,000,000+
Pluto TV - It’s Free TV : 1,000,000+
Mobile TV : 10,000,000+
TV+ : 5,000,000+
Digital TV : 5,000,000+
Motorola Spotlight Player™ : 10,000,000+
Vigo Lite : 5,000,000+
Hotstar : 100,000,000+
Peers.TV: broadcast TV channels First, Match TV, TNT ... : 5,000,000+
The green alien dance : 1,000,000+
Spectrum TV : 5,000,000+
H TV : 5,000,000+
StarTimes - Live International Champions Cup : 1,000,000+
Cinematic Cinematic : 1,000,000+
MEGOGO - Cinema and TV : 10,000,000+
Talking Angela : 100,000,000+
DStv Now : 5,000,000+
ivi - movies and TV shows in HD : 10,000,000+
Radio Javan : 1,000,000+
Talking Ginger 2 : 50,000,000+
Girly Lock Screen Wallpaper with Quotes : 5,000,000+
🔥 Football Wallpapers 4K | Full HD Backgrounds 😍 : 1,000,000+
Movies by Flixster, with Rotten Tomatoes : 10,000,000+
Low Poly – Puzzle art game : 1,000,000+
BBC Media Player : 10,000,000+
Amazon Prime Video : 50,000,000+
Adult Glitter Color by Number Book - Sandbox Pages : 1,000,000+
IMDb M

In [121]:
for apps in free_google_final_dataset:
    if apps[1] == 'GAME':
        print(apps[0], ":", apps[5])

Solitaire : 10,000,000+
Sonic Dash : 100,000,000+
PAC-MAN : 100,000,000+
Bubble Witch 3 Saga : 50,000,000+
Race the Traffic Moto : 10,000,000+
Marble - Temple Quest : 10,000,000+
Shooting King : 10,000,000+
Geometry Dash World : 10,000,000+
Jungle Marble Blast : 5,000,000+
Roll the Ball® - slide puzzle : 100,000,000+
Block Craft 3D: Building Simulator Games For Free : 50,000,000+
Farm Fruit Pop: Party Time : 1,000,000+
Love Balls : 50,000,000+
Piano Tiles 2™ : 100,000,000+
Pokémon GO : 100,000,000+
Paint Hit : 10,000,000+
Snake VS Block : 50,000,000+
Rolly Vortex : 10,000,000+
Woody Puzzle : 1,000,000+
Stack Jump : 10,000,000+
The Cube : 5,000,000+
Extreme Car Driving Simulator : 100,000,000+
Bricks n Balls : 1,000,000+
The Fish Master! : 1,000,000+
Color Road : 10,000,000+
Draw In : 10,000,000+
Block Craft 3D: Building Simulator Games For Free : 50,000,000+
PLANK! : 500,000+
Looper! : 1,000,000+
Trivia Crack : 100,000,000+
Will it Crush? : 5,000,000+
Tomb of the Mask : 5,000,000+
Base

Hence, `concluding`, the Entertainment and Games categories are more profitable scopes for android app development. The focus should be on Fun for the crowd to have more installations of their preferred apps. It shall be a profitable scope to invest into these sectors and later on grow with ios development too.